File: NestedListSum.java
// Time Complexity : O(N) N  is sum of count of integers and nested lists
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : NA


// Your code here along with comments explaining your approach

// We iterate through each nested integer in the given list
// If it is an integer we add n * d to sum
// If it is a list, we recursively calculate for the list with depth d + 1
// return sum

class Solution {
    
    int sum;
    
    public int depthSum(List<NestedInteger> nestedList) {
        
        if(nestedList == null) return sum;
        
        dfs(nestedList, 1);
        
        return sum;
    }
    
    private void dfs(List<NestedInteger> nestedList, int depth) {
        
        if(nestedList == null) return;
        
        for(NestedInteger el : nestedList) {
            if(el.isInteger()) {
                sum += el.getInteger() * depth;
            }
            else {
                dfs(el.getList(), depth + 1);
            }
        }
    }
}

File: nested_list_sum.py
# Approach - DFS 
# Time - O(N)
# Space - O(H) - height/depth of tree

class Solution:
    def depthSum(self, nestedList: List[NestedInteger]) -> int:
        
        return self.dfs(nestedList, depth=1)
        
        
    def dfs(self, nestedList, depth):
        total = 0
        
        for element in nestedList:
            if element.isInteger():
                total += element.getInteger() * depth
                
            else:
                total += self.dfs(element.getList(), depth + 1)
                
        return total



# BFS Approach
# O(N) Time and O(N) space

class Solution:
    def depthSum(self, nestedList: List[NestedInteger]) -> int:

        if not nestedList:
            return 0

        queue = deque()
        total = 0
        depth = 1

        for element in nestedList:
            queue.append(element)

        while queue:
            size = len(queue)
            
            for i in range(size):
                element = queue.popleft()

                if element.isInteger():
                    total += element.getInteger() * depth

                else:
                    for elem in element.getList():
                        queue.append(elem)

            depth += 1

        return total

